### DISCUSSION TITLE
Release Notes: v1.11.0

### DISCUSSION THREAD

**Discussion by PavelBlend** [PavelBlend/blender-xray#420] [General]
Здесь описаны нововведения в `blender-xray` версии `1.11.0`.

**Comment by PavelBlend** [comment]
Добавлены предупреждения в панель с параметрами импорта `omf`, если все опции отключены:
![01](https://user-images.githubusercontent.com/7983249/129493366-e7d8d518-57d6-48f7-b77e-92a00a8ef6d6.jpg)

**Comment by PavelBlend** [comment]
Добавлена поддержка экспорта `anm` анимаций, если режим вращения отличается от `YXZ`. В предыдущих версиях аддона происходила ошибка, если режим вращения не соответствовал режиму вращения, который используется в СДК:
![01](https://user-images.githubusercontent.com/7983249/129493435-220d5a40-d18d-42e3-8a67-d0d09eb2df3d.jpg)

Сейчас аддон в таких случаях будет делать запекание ключей и преобразование режима вращения.

**Comment by PavelBlend** [comment]
Добавлена поддержка конвертации `TCB` интерполяции в ключевые кадры при импорте `anm` файлов:
![00](https://user-images.githubusercontent.com/7983249/129493636-7e578ff2-bb94-417c-a026-a5c779c2efef.jpg)

В предыдущих версия до `v1.10.0` включительно, аддон импортировал TCB интерполяцию как Bezier ключи. Из-за этого анимация в блендере отличалась от анимации в игре.

В версии `v1.11.0` аддон делает преобразование TCB интерполяции в Linear.

**Reply by herm1t0** [reply]
С этим есть небольшая проблема, ключи проставляются некорректно, из-за этого весь результат искажается, появляются произвольные смещения в отдельных ключах, а так же слишком резкие переходы. Результат тот же даже в том случае, если я сам выставляю линейную интерполяцию в блендере и проставляю ключи на каждый кадр ручками

**Reply by PavelBlend** [reply]
@herm1t007 можешь скинуть пример файла? Я такой проблемы пока не видел.

**Reply by herm1t0** [reply]
@PavelBlend Конечно, вот примитивный пример
[Lean_right.zip](https://github.com/PavelBlend/blender-xray/files/7360528/Lean_right.zip)
В нем смещений нет, но это быстрый способ показать,что итоговый результат искажается и работать с ним в дальнейшем невозможно.

**Reply by PavelBlend** [reply]
@herm1t007 проверил файл. Я в аддоне ошибки не вижу, так как это такие ключи в файле сохранены. Как ты создал такой anm? В какой программе? В нём некоторые анимационные кривые имеют один ключ с типом STEPPED. А некоторые кривые имеют два ключа: первый STEPPED, а второй TCB. STEPPED ключи - это аналог Constant в блендере. Выглядит так:
![01](https://user-images.githubusercontent.com/7983249/137640176-7ef49e9c-6ee3-4974-845c-53bc40339dcf.jpg)

Если использовать STEPPED, то анимация будет скачкообразна. Нужно использовать LINEAR. Я импортировал твой файл и получил такие кривые:
![02](https://user-images.githubusercontent.com/7983249/137640275-db7f514b-f0d1-4571-b590-6af95c6638f3.jpg)

Я импортировал твой файл, но при импорте заменил STEPPED на LINEAR и получил такой результат:
![03](https://user-images.githubusercontent.com/7983249/137640321-9f41ff6a-4b8e-4362-a42a-667afc2d8d5a.jpg)

**Reply by herm1t0** [reply]
@PavelBlend Для создания использовал блендер. выставлял всего 2 TCB ключа на 0 и 12 кадрах. Остальные изменения уже произошли как раз после экспорта

**Reply by PavelBlend** [reply]
@herm1t007 это проблема экспорта. Блендер сохраняет bezier ключи как TCB. А вот почему первый ключ - это STEPPED? Должны все ключи быть TCB. У меня точно так же получилось, как в твоём файле. Я буду это выяснять. А пока включи запекание для своей анимации и сделай заново экспорт:
![1111111](https://user-images.githubusercontent.com/7983249/137642500-77c19250-6833-4a68-97b8-e0d97bf10653.jpg)

Если запекание установлено на режим On, то ключевые кадры будут установлены в каждом кадре и интерполяции TCB не будет.

**Reply by herm1t0** [reply]
@PavelBlend Спасибо большое :) Кстати, какую версию блендера предпочтительнее все же использовать?

**Reply by PavelBlend** [reply]
@herm1t007 я всегда последнюю использую. Сейчас установлена 2.93.

**Reply by PavelBlend** [reply]
@herm1t007 всё таки это была ошибка импорта. Проверял в игре anm файлы. Исправил преобразование интерполяции при импорте. Теперь аддон правильно импортирует. Фикс в последнем коммите.

Ошибка была в том, что я не правильно выбирал интерполяции ключей. В блендере интерполяция ключа влияет на форму кривой спереди:
![001](https://user-images.githubusercontent.com/7983249/137923768-35419a25-4f60-42aa-9384-5960be771f0c.jpg)

Но игра, сдк и 3д редактор lightwave (система интерполяций для сдк и движка была взята из него) используют другой алгоритм. Интерполяция ключа влияет на кривую сзади:
![02](https://user-images.githubusercontent.com/7983249/137924199-f84c10a0-f907-453f-b9eb-0073237b249d.jpg)


Этот баг проявлялся только тогда, когда первый ключ был STEPPED, а второй TCB. Я не встречал таких файлов. Обычно мне попадались anm и skls анимации с одним типом ключей вначале. Но эти файлы можно создать с помощью блендера. Но я об этом не знал. Блендер для первого ключа устанавливает STEPPED, так как интерполяция первого ключа ни на что не влияет.

**Comment by PavelBlend** [comment]
Добавлены пресеты для всех операторов импорта/экспорта:
![01](https://user-images.githubusercontent.com/7983249/129493721-21f436cf-a192-4346-a025-37a332108f6e.jpg)

**Comment by PavelBlend** [comment]
Добавлена поддержка автоматической установки путей, если первым параметром была указана папка `Objects Folder`:
![01](https://user-images.githubusercontent.com/7983249/129493759-fcdd673b-92f2-48e4-8525-b44c65753024.jpg)

**Comment by PavelBlend** [comment]
Изменена толщина линий в `blender` `2.8+` при отрисовке форм и лимитов костей.
Версия 1.10.0:
![02](https://user-images.githubusercontent.com/7983249/129493802-e29a13bc-8026-4b68-aa3b-6382fc21e893.jpg)
Версия 1.11.0:
![01](https://user-images.githubusercontent.com/7983249/129493809-7aa4a1a1-1f71-455b-bb17-4623e54a7ce2.jpg)
